cadc-datalink/src/test/java/org/opencadc/datalink/ServiceDescriptorTemplateTest.java

```diff
@@ -0,0 +1,228 @@
+        String template = Files.readString(testFile.toPath());
+
+        try {
+            new ServiceDescriptorTemplate(null, template);
+             Assert.fail("Expected IllegalArgumentException for null name");
+        } catch (IllegalArgumentException e) {
+            Assert.assertTrue(e.getMessage().contains("name cannot be null"));
+        }
+
+        try {
+            new ServiceDescriptorTemplate("", template);
+            Assert.fail("Expected IllegalArgumentException for empty name");
+        } catch (IllegalArgumentException e) {
+            Assert.assertTrue(e.getMessage().contains("name cannot be null"));
+        }
+
+        try {
+            new ServiceDescriptorTemplate("a-1?", template);
+            Assert.fail("Expected IllegalArgumentException for name with invalid character");
+        } catch (IllegalArgumentException e) {
+            Assert.assertTrue(e.getMessage().contains("Invalid descriptor name"));
+        }
+    }
+
+    @Test
+    public void testInvalidTemplateArgument() {
+
+        try {
+            ServiceDescriptorTemplate template = new ServiceDescriptorTemplate("test", null);
+            Assert.fail("Expected IllegalArgumentException for null template");
+        } catch (IllegalArgumentException e) {
+            Assert.assertTrue(e.getMessage().contains("template cannot be null"));
+        }
+
+        try {
+            ServiceDescriptorTemplate template = new ServiceDescriptorTemplate("test", "");
+            Assert.fail("Expected IllegalArgumentException for empty template");
+        } catch (IllegalArgumentException e) {
+            Assert.assertTrue(e.getMessage().contains("template cannot be null"));
+        }
+
+        try {
+            ServiceDescriptorTemplate template = new ServiceDescriptorTemplate("test", "template");
+            Assert.fail("Expected IllegalArgumentException for template not a votable");
+        } catch (IllegalArgumentException e) {
+            Assert.assertTrue(e.getMessage().contains("Error reading VOTable"));
+        }
+
+        try {
+            ServiceDescriptorTemplate template = new ServiceDescriptorTemplate("test", "<foo></foo>");
+            Assert.fail("Expected IllegalArgumentException for template an empty votable");
+        } catch (IllegalArgumentException e) {
+            Assert.assertTrue(e.getMessage().contains("Error reading VOTable"));
+        }
+
+    }
+
+    @Test
+    public void testValidTemplate() throws Exception {
+
+        File testFile = FileUtil.getFileFromResource("valid-template.xml", ServiceDescriptorTemplateTest.class);
+        String template = Files.readString(testFile.toPath());
+
+        try {
+            new ServiceDescriptorTemplate("test", template);
+        } catch (IllegalArgumentException e) {
+            Assert.fail("Expected no exception for valid template");
+        }
+    }
+
+    @Test
+    public void testInvalidIdRefTemplate() throws Exception {
+
```

Comment:
All of these fail because of the exception from the `read` method; that error message is from the XML parser.

So most of these tests are just more elaborate than necessary and the code to re-validate never actually does anything useful. That's kind of the point of using xml: a lot of validation is built in (sure, error messages can be bad but that's a different problem). One test with a very simple mismatched ID/IDREF usage is sufficient, so probably just keep the mismatched-id-ref-template but remove the matching one and just have a simple mismatched one.

I'll make some comments in the valid-template about minimal content in there, but it should apply to all the test files.

Reply:
Cleaned up the unit tests to check only what is thrown outside of the schema validation.